ADAL/src/request/ADAuthenticationRequest+AcquireToken.m

```diff
@@ -94,135 +94,232 @@ - (void)validatedAcquireToken:(ADAuthenticationCallback)completionBlock
-    }
-    
     if (![ADAuthenticationContext isForcedAuthorization:_promptBehavior] && [_context hasCacheStore])
     {
-        //Cache should be used in this case:
-        ADTokenCacheItem* cacheItem = [self findCacheItemWithKey:key
-                                                          userId:_identifier
-                                                           error:&error];
-        if (!cacheItem && error)
-        {
-            completionBlock([ADAuthenticationResult resultFromError:error correlationId:_correlationId]);
-            return;
-        }
-        
-        if (cacheItem)
-        {
-            //Found a promising item in the cache, try using it:
-            [self attemptToUseCacheItem:cacheItem completionBlock:completionBlock];
-            return; //The tryRefreshingFromCacheItem has taken care of the token obtaining
-        }
+        [self tryAccessToken:completionBlock];
+        return;
     }
     
     [self requestToken:completionBlock];
 }
 
+- (void)acquireTokenWithItem:(ADTokenCacheItem *)item
+                 refreshType:(NSString *)refreshType
+             completionBlock:(ADAuthenticationCallback)completionBlock
+                    fallback:(ADAuthenticationCallback)fallback
+{
+    [self acquireTokenByRefreshToken:item.refreshToken
+                           cacheItem:item
+                     completionBlock:^(ADAuthenticationResult *result)
+     {
+         NSString* msg = nil;
+         if (refreshType)
+         {
+             msg = [NSString stringWithFormat:@"Acquire Token with %@ Refresh Token %@.", refreshType, result.status == AD_SUCCEEDED ? @"Succeeded" : @"Failed"];
+         }
+         else
+         {
+             msg = [NSString stringWithFormat:@"Acquire Token with Refresh Token %@.", result.status == AD_SUCCEEDED ? @"Succeeded" : @"Failed"];
+         }
+         
+         AD_LOG_INFO_F(msg, _correlationId, @"clientId: '%@'; resource: '%@';", _clientId, _resource);
+         
+         if ([ADAuthenticationContext isFinalResult:result])
+         {
+             completionBlock(result);
+             return;
+         }
+         
+         fallback(result);
+     }];
+}
+
 /*Attemps to use the cache. Returns YES if an attempt was successful or if an
  internal asynchronous call will proceed the processing. */
-- (void)attemptToUseCacheItem:(ADTokenCacheItem*)item
-              completionBlock:(ADAuthenticationCallback)completionBlock
+- (void)tryAccessToken:(ADAuthenticationCallback)completionBlock
 {
     //All of these should be set before calling this method:
     THROW_ON_NIL_ARGUMENT(completionBlock);
-    AD_REQUEST_CHECK_ARGUMENT(item);
     AD_REQUEST_CHECK_PROPERTY(_resource);
     AD_REQUEST_CHECK_PROPERTY(_clientId);
     
     [self ensureRequest];
     
+    BOOL fADFSUser = NO;
+    
+    ADAuthenticationError* error = nil;
+    ADTokenCacheItem* item = [self getItemForResource:_resource clientId:_clientId error:&error];
+    if (!item && error)
```

Comment:
Add some comments around the if checks so that someone new can understand what is going on.